### PR TITLE
SI-5568 Fixes verify error from getClass on refinement of value type

### PR DESCRIPTION
**Pull request by JamesIry**
().asInstanceOf[AnyRef with Unit].getClass and
5.asInstanceOf[AnyRef with Int].getClass would cause a verify
error. Going the other way, i.e. [Unit with AnyRef] or [Int with AnyRef]
worked fine. This commit fixes it that both directions work out to
BoxedUnit or java.lang.Integer.

review @gkossakowski

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 1]
```diff
@@ -1152,9 +1158,14 @@ abstract class Erasure extends AddInterfaces
               } else if (isPrimitiveValueClass(qual.tpe.typeSymbol)) {
                 // Rewrite 5.getClass to ScalaRunTime.anyValClass(5)
                 global.typer.typed(gen.mkRuntimeCall(nme.anyValClass, List(qual, typer.resolveClassTag(tree.pos, qual.tpe.widen))))
-              } else if (fn.symbol == AnyVal_getClass) {
+              } else if (primitiveGetClassMethods.contains(fn.symbol)) {
+                // if we got here then we're trying to send a primitive getClass method to 
+                // a) an Any, or
+                // b) a non-primitive, e.g. because the qualifier's type is a refinement type where part of the refinement is a primitive.
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: `part` is probably not the best word. Did you mean `parent`?

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 2]
```diff
@@ -340,12 +340,18 @@ abstract class Erasure extends AddInterfaces
       case _                               => tp.deconst
     }
   }
-  // Methods on Any/Object which we rewrite here while we still know what
-  // is a primitive and what arrived boxed.
-  private lazy val interceptedMethods = Set[Symbol](Any_##, Object_##, Any_getClass, AnyVal_getClass) ++ (
-    // Each value class has its own getClass for ultra-precise class object typing.
+  
+  // Each value class has its own getClass for ultra-precise class object typing.
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: I guess we need a more precise term for "value class" now that `AnyVal` is no longer sealed. "Primitive value class"?

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 3]
```diff
@@ -340,12 +340,18 @@ abstract class Erasure extends AddInterfaces
...
+  private lazy val primitiveGetClassMethods = Set[Symbol](Any_getClass, AnyVal_getClass) ++ (
     ScalaValueClasses map (_.tpe member nme.getClass_)
   )
+  
+  // ## requires a little translation
+  private lazy val poundPoundMethods = Set[Symbol](Any_##, Object_##)
+  
+  // Methods on Any/Object which we rewrite here while we still know what
+  // is a primitive and what arrived boxed.
+  private lazy val interceptedMethods = poundPoundMethods ++ primitiveGetClassMethods
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:354]: Thanks for cleaning this up!

File: src/compiler/scala/tools/nsc/transform/Erasure.scala (hunk 2 above)
Comment (JamesIry) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: I'm good with that

File: src/compiler/scala/tools/nsc/transform/Erasure.scala (hunk 1 above)
Comment (JamesIry) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: parent is fine

**Comment by gkossakowski** [comment]
@JamesIry, it would be useful to add a comment in commit message why boxing is now properly triggered (it's handled later in erasure phase). I know you mention that on a ticket but it's easier to use tools like git blame if essential info (conclusion) is in commit message. Please do that only if you are about to change something else (e.g change comments in code).

Other than that LGTM!

**Comment by JamesIry** [comment]
I can beef up that comment while I'm working on the other comment bits.

**Comment by JamesIry** [comment]
ping @gkossakowski for final review

**Comment by gkossakowski** [comment]
@JamesIry: I didn't notice your new commit. Everything Looks Good to Me now.

**Pull request merged**